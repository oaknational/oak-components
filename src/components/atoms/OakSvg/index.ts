export { OakSvg } from "./OakSvg";
export type { OakSvgProps, OakSvgNames } from "./OakSvg";
