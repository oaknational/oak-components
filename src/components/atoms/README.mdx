# Atoms

## Rules for components in this folder

- Can import other atoms
- Imports from molecules and organisms are forbidden!
- Generic
- Unstyled
- Avoid embedded logic and state
- No sub-components
  (eg. Box)
