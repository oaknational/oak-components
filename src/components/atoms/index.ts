export * from "./OakBox";
export * from "./OakFlex";
export * from "./OakSpan";
export * from "./OakForm";
export * from "./OakImage";
export * from "./OakTypography";
export * from "./OakHeading";
export * from "./OakP";
export * from "./OakUL";
export * from "./OakOL";
export * from "./OakLI";
export * from "./OakLabel";
export * from "./OakIcon";
export * from "./OakGrid";
export * from "./OakGridArea";
export * from "./OakThemeProvider";
export * from "./OakMaxWidth";
export * from "./OakCloudinaryImage";
export * from "./OakKbd";
export * from "./OakGlobalStyle";
export * from "./OakAnchorTarget";
export * from "./OakFieldSet";
