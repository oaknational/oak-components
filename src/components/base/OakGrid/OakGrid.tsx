import styled, { css } from "styled-components";

import {
  responsiveStyle,
  ResponsiveValues,
} from "@/styles/utils/responsiveStyle";
import { OakBox, OakBoxProps } from "@/components/base/OakBox";
import { parseSpacing } from "@/styles/helpers/parseSpacing";
import { OakCombinedSpacingToken } from "@/styles";

const gridStyle = css<OakGridProps>`
  ${responsiveStyle("row-gap", (props) => props.$rg, parseSpacing)}
  ${responsiveStyle("column-gap", (props) => props.$cg, parseSpacing)}
  ${responsiveStyle("grid-auto-rows", (props) => props.$gridAutoRows)}
  ${responsiveStyle("grid-template-areas", (props) => props.$gridTemplateAreas)}
  ${responsiveStyle(
    "grid-template-columns",
    (props) => props.$gridTemplateColumns,
  )}
`;

export type OakGridProps = OakBoxProps & {
  $rg?: ResponsiveValues<OakCombinedSpacingToken>;
  $cg?: ResponsiveValues<OakCombinedSpacingToken>;
  // grid-auto-rows: 1fr; ensures all rows are the same height
  $gridAutoRows?: ResponsiveValues<"1fr">;
  $gridTemplateAreas?: ResponsiveValues<string>;
  $gridTemplateColumns?: ResponsiveValues<string>;
};

export const OakGrid = styled(OakBox)<OakGridProps>`
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  width: 100%;
  ${gridStyle}
`;
