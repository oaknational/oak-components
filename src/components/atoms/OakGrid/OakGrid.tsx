import styled, { css, CSSProperties } from "styled-components";

import {
  responsiveStyle,
  ResponsiveValues,
} from "@/styles/utils/responsiveStyle";
import { OakBox, OakBoxProps } from "@/components/atoms/OakBox";
import { parseSpacing } from "@/styles/helpers/parseSpacing";
import { OakCombinedSpacingToken } from "@/styles";

const gridStyle = css<OakGridProps>`
  ${responsiveStyle("row-gap", (props) => props.$rg, parseSpacing)}
  ${responsiveStyle("column-gap", (props) => props.$cg, parseSpacing)}
  ${responsiveStyle("grid-auto-rows", (props) => props.$gridAutoRows)}
  ${responsiveStyle("grid-template-areas", (props) => props.$gridTemplateAreas)}
  ${responsiveStyle(
    "grid-template-columns",
    (props) => props.$gridTemplateColumns,
  )}
    ${responsiveStyle("grid-template-rows", (props) => props.$gridTemplateRows)}
`;

export type OakGridProps = OakBoxProps & {
  /**
   * Applies `row-gap` to the grid
   *
   * Accepts a spacing token or a responsive array of spacing tokens.
   */
  $rg?: ResponsiveValues<OakCombinedSpacingToken>;
  /**
   * Applies `column-gap` to the grid
   *
   * Accepts a spacing token or a responsive array of spacing tokens.
   */
  $cg?: ResponsiveValues<OakCombinedSpacingToken>;
  /**
   * Applies `grid-auto-rows` to the grid
   *
   * Accepts a `grid-auto-rows` value or a responsive array of `grid-auto-rows` values.
   */
  $gridAutoRows?: ResponsiveValues<CSSProperties["gridAutoRows"]>;
  /**
   * Applies `grid-template-areas` to the grid
   *
   * Accepts a `grid-template-areas` value or a responsive array of `grid-template-areas` values.
   */
  $gridTemplateAreas?: ResponsiveValues<CSSProperties["gridTemplateAreas"]>;
  /**
   * Applies `grid-template-columns` to the grid
   *
   * Accepts a `grid-template-columns` value or a responsive array of `grid-template-columns` values.
   */
  $gridTemplateColumns?: ResponsiveValues<CSSProperties["gridTemplateColumns"]>;
  /**
   * Applies `grid-template-rows` to the grid
   *
   * Accepts a `grid-template-rows` value or a responsive array of `grid-template-rows` values.
   */
  $gridTemplateRows?: ResponsiveValues<CSSProperties["gridTemplateRows"]>;
};

/**
 * Creates a grid layout
 *
 * Defaults to a 12 column grid
 */
export const OakGrid = styled(OakBox)<OakGridProps>`
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  width: 100%;

  ${gridStyle};
`;
