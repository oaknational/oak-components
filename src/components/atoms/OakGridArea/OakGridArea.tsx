import styled, { css, CSSProperties } from "styled-components";

import {
  responsiveStyle,
  ResponsiveValues,
} from "@/styles/utils/responsiveStyle";
import { SpacingStyleProps } from "@/styles/utils/spacingStyle";
import { OakFlex, OakFlexProps } from "@/components/atoms/OakFlex";

type ColRowSpan = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export type OakGridAreaProps = {
  /**
   * Determines the number of columns the element should span.
   *
   * Accepts a value from 0-12 or a responsive value of 0-12.
   */
  $colSpan: ResponsiveValues<ColRowSpan>;
  /**
   * Determines the number of rows the element should span.
   *
   * Accepts a value from 0-12 or a responsive value of 0-12.
   */
  $rowSpan?: ResponsiveValues<ColRowSpan>;
  /**
   * Sets the order of the element.
   *
   * Accepts a number or a responsive array of numbers.
   */
  $order?: ResponsiveValues<CSSProperties["order"]>;
  /**
   * The start column of the element.
   *
   * Accepts a value from 0-12 or a responsive value of 0-12.
   */
  $colStart?: ResponsiveValues<ColRowSpan>;
  /**
   * The end column of the element.
   *
   * Accepts a value from 0-12 or a responsive value of 0-12.
   */
  $colEnd?: ResponsiveValues<ColRowSpan>;
  /**
   * The start row of the element.
   *
   * Accepts a value from 0-12 or a responsive value of 0-12.
   */
  $rowStart?: ResponsiveValues<ColRowSpan>;
} & SpacingStyleProps;

const combineSpanStart = (
  start?: ColRowSpan | null,
  span?: ColRowSpan | null,
) => {
  return start ? `${start}/${span}` : `${span}`;
};

const parseSpanStart = (value?: string | null) => {
  if (value?.includes("/")) {
    const [start, span] = value.split("/");
    return `${start} / span ${span}`;
  }
  const span = value;
  return `span ${span}`;
};

const gridArea = css<OakGridAreaProps>`
  flex-direction: column;
  ${responsiveStyle(
    "grid-column",
    (props) => {
      return Array.isArray(props.$colSpan)
        ? props.$colSpan.map((span, index) =>
            combineSpanStart(
              Array.isArray(props.$colStart)
                ? props.$colStart[index]
                : props.$colStart,
              span,
            ),
          )
        : combineSpanStart(
            Array.isArray(props.$colStart)
              ? props.$colStart[0]
              : props.$colStart,
            props.$colSpan,
          );
    },
    (value) => parseSpanStart(value),
  )};
  ${responsiveStyle(
    "grid-row",
    (props) => {
      return Array.isArray(props.$rowSpan)
        ? props.$rowSpan.map((span, index) =>
            combineSpanStart(
              Array.isArray(props.$rowStart)
                ? props.$rowStart[index]
                : props.$rowStart,
              span,
            ),
          )
        : combineSpanStart(
            Array.isArray(props.$rowStart)
              ? props.$rowStart[0]
              : props.$rowStart,
            props.$rowSpan,
          );
    },
    (value) => parseSpanStart(value),
  )};
  ${responsiveStyle(
    "order",
    (props) => props.$order,
    (value) => value && `${value}`,
  )};
  ${responsiveStyle("grid-row", (props) =>
    props.$rowSpan ? `span ${props.$rowSpan}` : "span 1",
  )};
  ${responsiveStyle("grid-column-start", (props) => props.$colStart)}
  ${responsiveStyle("grid-column-end", (props) => props.$colEnd)}
  ${responsiveStyle("grid-row-start", (props) => props.$rowStart)}
`;

/**
 *
 * OakGridArea is a flex container that should be used inside OakGrid.
 *
 * - There is no nesting of OakGridAreas
 * - Column and row arrangements are achieved through $colSpan, $rowSpan, $colStart, $colEnd, $rowStart, $rowEnd
 *
 */

export const OakGridArea = styled(OakFlex)<OakGridAreaProps & OakFlexProps>`
  ${gridArea}
`;
