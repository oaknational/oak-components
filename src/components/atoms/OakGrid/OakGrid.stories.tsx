import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakGrid } from "./OakGrid";

import { OakGridArea } from "@/components/atoms/OakGridArea";
import { marginCtl } from "@/storybook-helpers/spacingStyleHelpers";

const meta: Meta<typeof OakGrid> = {
  component: OakGrid,
  tags: ["autodocs"],
  title: "components/atoms/OakGrid",
  argTypes: {
    $rg: marginCtl,
    $cg: marginCtl,
    $gridAutoRows: {
      control: "text",
    },
    $gridTemplateAreas: {
      control: "text",
    },
    $gridTemplateColumns: {
      control: "text",
    },
    $gridTemplateRows: {
      control: "text",
    },
  },
  parameters: {
    controls: {
      include: [
        "$rg",
        "$cg",
        "$gridAutoRows",
        "$gridTemplateAreas",
        "$gridTemplateColumns",
        "$gridTemplateRows",
      ],
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakGrid>;

export const Default: Story = {
  render: (args) => (
    <OakGrid {...args}>
      <OakGridArea $background={"amber"} $colSpan={6} $pa="spacing-24">
        6 columns
      </OakGridArea>
      <OakGridArea $background={"mint"} $colSpan={4} $pa="spacing-24">
        4 columns
      </OakGridArea>
      <OakGridArea $background={"lemon"} $colSpan={2} $pa="spacing-24">
        2 columns
      </OakGridArea>
      <OakGridArea $background={"teal"} $colSpan={12} $pa="spacing-24">
        12 columns
      </OakGridArea>
    </OakGrid>
  ),
  args: {
    $rg: "spacing-24",
    $cg: "spacing-24",
  },
};
