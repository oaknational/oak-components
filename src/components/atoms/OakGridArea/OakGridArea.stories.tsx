import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakGridArea } from "./OakGridArea";

import { OakGrid } from "@/components/atoms/OakGrid";

const colRowSpanCtl = { control: { type: "range" as const, min: 0, max: 12 } };

const meta: Meta<typeof OakGridArea> = {
  component: OakGridArea,
  tags: ["autodocs"],
  title: "components/atoms/OakGridArea",
  argTypes: {
    $colSpan: colRowSpanCtl,
    $rowSpan: colRowSpanCtl,
    $order: colRowSpanCtl,
    $colStart: colRowSpanCtl,
    $colEnd: colRowSpanCtl,
    $rowStart: colRowSpanCtl,
  },
  parameters: {
    docs: {
      inlineStories: false,
    },
    controls: {
      include: [
        "$colSpan",
        "$rowSpan",
        "$order",
        "$colStart",
        "$colEnd",
        "$rowStart",
      ],
    },
  },
};

export default meta;

type Story = StoryObj<typeof OakGridArea>;

export const SpanningColumns: Story = {
  render: (args) => (
    <OakGrid>
      <OakGridArea
        {...args}
        $background={"amber"}
        $colSpan={6}
        $pa="spacing-24"
      >
        6 columns
      </OakGridArea>
      <OakGridArea $background={"mint"} $colSpan={4} $pa="spacing-24">
        4 columns
      </OakGridArea>
      <OakGridArea $background={"lemon"} $colSpan={2} $pa="spacing-24">
        2 columns
      </OakGridArea>
      <OakGridArea $background={"teal"} $colSpan={12} $pa="spacing-24">
        12 columns
      </OakGridArea>
    </OakGrid>
  ),
  parameters: { controls: { include: [] } },
};

export const SpanningRows: Story = {
  render: (args) => (
    <OakGrid>
      <OakGridArea
        {...args}
        $background={"amber"}
        $colSpan={6}
        $rowSpan={2}
        $pa="spacing-24"
      >
        6 columns, 2 rows
      </OakGridArea>
      <OakGridArea $background={"mint"} $colSpan={6} $pa="spacing-24">
        6 columns, 1 row
      </OakGridArea>
      <OakGridArea $background={"lemon"} $colSpan={3} $pa="spacing-24">
        3 columns, 1 row
      </OakGridArea>
      <OakGridArea $background={"teal"} $colSpan={3} $pa="spacing-24">
        3 columns, 1 row
      </OakGridArea>
    </OakGrid>
  ),
  parameters: { controls: { include: [] } },
};

export const StartCol: Story = {
  render: (args) => (
    <OakGrid>
      <OakGridArea
        {...args}
        $background={"amber"}
        $colSpan={3}
        $pa="spacing-24"
        $colStart={3}
      >
        Col start 3
      </OakGridArea>
      <OakGridArea
        $background={"lemon"}
        $colSpan={3}
        $colStart={4}
        $pa="spacing-24"
      >
        Col start 4
      </OakGridArea>
      <OakGridArea
        $background={"teal"}
        $colSpan={3}
        $colStart={5}
        $pa="spacing-24"
      >
        Col Start 5
      </OakGridArea>
    </OakGrid>
  ),
  parameters: { controls: { include: [] } },
};

export const EndCol: Story = {
  render: (args) => (
    <OakGrid>
      <OakGridArea
        {...args}
        $background={"amber"}
        $colSpan={3}
        $pa="spacing-24"
        $colEnd={4}
      >
        Col end 4
      </OakGridArea>
      <OakGridArea
        $background={"lemon"}
        $colSpan={3}
        $colEnd={5}
        $pa="spacing-24"
      >
        Col end 5
      </OakGridArea>
      <OakGridArea
        $background={"teal"}
        $colSpan={3}
        $colEnd={6}
        $pa="spacing-24"
      >
        Col end 6
      </OakGridArea>
    </OakGrid>
  ),
  parameters: { controls: { include: [] } },
};
