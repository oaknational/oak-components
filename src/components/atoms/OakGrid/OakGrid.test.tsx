import React from "react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { create } from "react-test-renderer";

import { OakGrid } from "./OakGrid";

import { OakGridArea } from "@/components/atoms/OakGridArea";

describe("OakGrid", () => {
  it("renders without crashing", () => {
    const { getByTestId } = render(
      <OakGrid data-testid="oak-grid" $cg={"all-spacing-1"} />,
    );
    expect(getByTestId("oak-grid")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = create(<OakGrid />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("is a grid", () => {
    const { getByTestId } = render(<OakGrid data-testid="oak-grid" />);
    expect(getByTestId("oak-grid")).toHaveStyle("display: grid;");
  });

  it("has 12 columns", () => {
    const { getByTestId } = render(<OakGrid data-testid="oak-grid" />);
    expect(getByTestId("oak-grid")).toHaveStyle(
      "grid-template-columns: repeat(12, 1fr);",
    );
  });

  it("adjusts column gap via $cg", () => {
    const { getByTestId } = render(
      <OakGrid data-testid="oak-grid" $cg={"all-spacing-1"}>
        <OakGridArea $colSpan={1} />
      </OakGrid>,
    );
    expect(getByTestId("oak-grid")).toHaveStyle("column-gap: 0.25rem;");
  });
});
