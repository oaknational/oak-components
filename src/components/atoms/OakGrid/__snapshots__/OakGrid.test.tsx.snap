// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakGrid matches snapshot 1`] = `
.c0:where(span) {
  display: block;
}

.c0:where(*) {
  font-family: --var(google-font),Lexend,sans-serif;
}

.c1 {
  display: grid;
  grid-template-columns: repeat(12,1fr);
  width: 100%;
}

<div
  className="c0 c1"
/>
`;
